### PR TITLE
Add new `paredit.[backward/forwardSexpOrUp]`. Fixes #1657

### LINKED ISSUE #1657: Combined command for `backward/forwardSexp` and `backward/forwardUpSexp` - `backward/forwardSexpOrUp`
As per some [discussion on slack](https://clojurians.slack.com/archives/CBE668G4R/p1648707972296699), I'd like to request a new paredit command that will go up if at the end of an sexp. 

Current `forwardSexp` behavior: (`|` is cursor, each line is an invocation of fwdSexp)
 ```clojure
(def x (|inc 1))
(def x (inc |1))
(def x (inc 1|))
(def x (inc 1|)) ;; stopped moving
```

To more closely match some paredit behavior from Cursive that I use, it'd be nice to have this modified behavior:

 ```clojure
(def x (|inc 1))
(def x (inc |1))
(def x (inc 1|))
(def x (inc 1)|) ;; Escaped the sexp
(def x (inc 1))| ;; Escaped the sexp
```

### PR DESCRIPTION
**Pull request by JJ-Atkinson**
## What has Changed?

I've tried to make the minimal change necessary to implement the new paredit commands. I didn't see any tests for the orignal commands, and there's a note in the contrib guide about the challenges testing the token cursor, so I've not written any new ones. 

Fixes #1657

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
     - [x] If I am fixing just part of the issue, I have just referenced it w/o any of the "fixes” keywords.
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping pez, bpringe, corasaurus-hex, Cyrik

### CODE REVIEW

**Comment by JJ-Atkinson** [comment]
Remaining things to do:

- [x] Add default keybindings if required to these new commands. I don't really know what to suggest.
- [x] Add GIFS that match the current style into the documentation
- [x] (maybe) Add `select` versions of `...SexpOrUp` for consistency

File: docs/site/paredit.md [hunk 1]
```diff
@@ -64,6 +64,8 @@ Default keybinding      | Action | Description
  `ctrl+alt+up`             | **Backward Down Sexp** | Moves the cursor into the preceding list.<br> ![](images/paredit/backward-down-sexp.gif)
  `ctrl+alt+down`           | **Forward Up Sexp** | Moves the cursor forwards, out of the current list.<br> ![](images/paredit/forward-up-sexp.gif)
  `ctrl+up`                 | **Backward Up Sexp** | Moves the cursor backwards, out of the current list.<br> ![](images/paredit/backward-up-sexp.gif)
+ TODO                      | **Forward Sexp Or Up** | Moves the cursor forward, to the end of the current form. If at the end, moves to the end of the next form. Moves out of lists if at the end. <br> ![](TODO)
+ TODO                      | **Backward Sexp Or Up** | Moves the cursor backward, to the start of the current form. If at the start, moves to the start of the previous form. Moves out of lists if at the end. <br> ![](TODO)
```
Comment (PEZ) [inline docs/site/paredit.md]: I think it makes sense to leave these unbound by default.

**Comment by PEZ** [comment]
Thanks so much for this work!!!

I think we can leave the commands default without keybindings. I'm hoping that the way you've added it now without keys, will help the users get the when context right. Will tests this.

For the GIFs I've used Solarized light.

I agree about `select` versions. It's sort of a contract, as I see it. 😄 

Let's figure if we can have less duplication. Not sure what is the best approach...

Please add some tests in `src/extension-test/unit/cursor-doc/paredit-test.ts`

File: package.json [hunk 2]
```diff
@@ -1888,6 +1900,14 @@
...
         "when": "calva:keybindingsEnabled && editorLangId == clojure && editorTextFocus && paredit:keyMap =~ /original|strict/ && config.calva.paredit.hijackVSCodeDefaults && !calva:cursorInComment || calva:cursorAfterComment"
       },
+      {
+        "command": "paredit.forwardSexpOrUp",
+        "when": "calva:keybindingsEnabled && editorLangId == clojure && editorTextFocus && paredit:keyMap =~ /original|strict/ && config.calva.paredit.hijackVSCodeDefaults && !calva:cursorInComment || calva:cursorAfterComment"
+      },
+      {
+        "command": "paredit.backwardSexpOrUp",
+        "when": "calva:keybindingsEnabled && editorLangId == clojure && editorTextFocus && paredit:keyMap =~ /original|strict/ && config.calva.paredit.hijackVSCodeDefaults && !calva:cursorInComment || calva:cursorAfterComment"
+      },
```
Comment (PEZ) [inline package.json]: I tested this now. The when expression does not automatically appear in the shortcut editor. So then it is probably better to just skip this.

Comment (JJ-Atkinson) [reply package.json]: What do you mean by that? I see it in the command pallet if I search

Comment (PEZ) [reply package.json]: Yes, but this is key binding config, right? Just lacking the actual key. Unless I misread something. I was hoping that when you open the keyboard shortcuts editor, then the when context from this config would be filled in when you added a shortcut for one of these commands. But it doesn't. So then these configs are just noise.

Comment (JJ-Atkinson) [reply package.json]: Ah, I misunderstood the purpose of this then. I thought it was required for plugin exports

**Comment by JJ-Atkinson** [comment]
@PEZ I think we've got it sorted now. I extracted the `forwardSexp`, `forwardUpSexp` logic into a common fn. I left the tests more or less as is - there are a couple that test moving up a level. I don't know enough about TS to know if you can/should test non exported things.

**Review by PEZ — APPROVED** [review]
Thanks! 🙏 ❤️

**Comment by PEZ** [comment]
Testing non-exported isn't even possible. 😄 

But wether to test the function is question here is not so much about TS. If you would test it in Clojure you would test it in TS. I think it might make sense to test it, but am also fine with having it tested indirectly for now.

**Pull request merged**